pypdf/_text_extraction/_layout_mode/_fonts.py

```diff
@@ -0,0 +1,486 @@
+        displaced_tx (float): tx from displaced_xform
+        space_tx (float): tx for a space character
+        font_height (float): font height
+
+    Methods:
+        font_size_matrix: matrix accounting for font size, horizontal scale, and text rise.
+        displacement_matrix: matrix accounting for horizontal translation during rendering.
+    """
+
+    txt: str
+    font: Font
+    font_size: Union[int, float]
+    Tc: float = 0.0
+    Tw: float = 0.0
+    Tz: float = 100.0
+    TL: float = 0.0
+    Ts: float = 0.0
+    xform: List[float] = field(default_factory=lambda: [1.0, 0.0, 0.0, 1.0, 0.0, 0.0])
+    displaced_xform: List[float] = field(
+        default_factory=lambda: [1.0, 0.0, 0.0, 1.0, 0.0, 0.0], init=False
+    )
+    render_xform: List[float] = field(
+        default_factory=lambda: [1.0, 0.0, 0.0, 1.0, 0.0, 0.0], init=False
+    )
+    displaced_tx: float = field(default=0.0, init=False)
+    space_tx: float = field(default=0.0, init=False)
+    font_height: float = field(default=0.0, init=False)
+    flip_vertical: bool = field(default=False, init=False)
+
+    def __post_init__(self) -> None:
+        self.displaced_xform = mult(self.displacement_matrix(), self.xform)
+        self.render_xform = mult(self.font_size_matrix(), self.xform)
+        self.displaced_tx = self.displaced_xform[4]
+        self.space_tx = round(
+            mult(self.displacement_matrix(" "), self.xform)[4] - self.xform[4], 3
+        )
+        self.font_height = self.font_size * math.sqrt(self.xform[1] ** 2 + self.xform[3] ** 2)
+        self.flip_vertical = self.xform[3] < -1e-6
+
+    def font_size_matrix(self) -> List[float]:
+        """Font size matrix"""
+        return [self.font_size * (self.Tz / 100.0), 0.0, 0.0, self.font_size, 0.0, self.Ts]
+
+    def displacement_matrix(self, word: Union[str, None] = None, TD_offset: float = 0.0) -> List[float]:
+        """
+        Text displacement matrix
+
+        Args:
+            word (str, optional): Defaults to None in which case self.txt displacement is
+                returned.
+            TD_offset (float, optional): translation applied by TD operator. Defaults to 0.0.
+        """
+        word = word or self.txt
+        return [1.0, 0.0, 0.0, 1.0, self.word_tx(word, TD_offset), 0.0]
+
+    def word_tx(self, word: str, TD_offset: float = 0.0) -> float:
+        """Horizontal text displacement for any word according this text state"""
+        return (
+            (self.font_size * ((self.font.word_width(word) - TD_offset) / 1000.0))
+            + self.Tc
+            + self.txt.count(" ") * self.Tw
+        ) * (self.Tz / 100.0)
+
+    @staticmethod
+    def to_dict(inst: "TextStateParams") -> Dict[str, Any]:
+        """Dataclass to dict for json.dumps serialization"""
+        return {k: getattr(inst, k) for k in inst.__dataclass_fields__ if k != "font"}
+
+
+# Widths for some of the standard 14 fonts
```

Comment:
Where did you get them from? Would it make sense to make this a subpackage and every of the fonts a single file (maybe even YAML) so that it's clearer how to extend this?

Reply:
As for my original source (and I'm sure you'll love this), it was ChatGPT[^1], but I verified the values with a combination of googling, examination of local OS font files, and manual testing. In any case, the true character widths aren't as important as you might think. A weighted average fixed character width is calculated across the displacements of all text render operations[^2], and it's this fixed character width in combination with the starting x position of each text render operator's transformation matrix[^3] that's used to estimate how many spaces should be inserted to preserve horizontal positioning. Averaging the fixed width across the render operations for the entire page "smooths out" most errors in the original width map.

Re extensibility, I don't think we'll need it, at least not for the sake of widths. 12 of the 14 standard fonts were covered in the initial commit (4 variants each for Times, Courier, and Helvetica). I've since added mappings for ZapfDingbats and Symbol, and provided aliasing according to table H.3 on page 1110 of the pdf 1.7 standard. With that in place, I'm hoping this particular use case is set for the long term. Encoding concerns (e.g. predefined CMaps for CJK fonts) are another matter, but I think that's OOS for this PR?

[^1]: The [chat](https://chat.openai.com/share/249334cb-7545-4035-9e3e-2628fb9c5471) where I obtained the standard width values.
[^2]: Technically, all text render operations between BT/ET pairs are aggregated into a BTGroup, and the average is taken across the set of all BTGroups.
[^3]: Again, the starting positions are actually those of BTGroups, not individual render operations.